python/lsst/ip/diffim/metrics.py

```diff
@@ -0,0 +1,158 @@
+
+            ``measurement``
+                the total number of science sources (`lsst.verify.Measurement`
+                or `None`)
+        """
+        nSciSources = 0
+        inputData = False
+        for catalog in sources:
+            if catalog is not None:
+                nSciSources += len(catalog)
+                inputData = True
+
+        if inputData:
+            meas = Measurement(self.getOutputMetricName(self.config), nSciSources * u.count)
+        else:
+            self.log.info("Nothing to do: no catalogs found.")
+            meas = None
+        return Struct(measurement=meas)
+
+    @classmethod
+    def getOutputMetricName(cls, config):
+        return "ip_diffim.numSciSources"
+
+
+class FractionDiaSourcesToSciSourcesMetricConfig(MetricTask.ConfigClass):
+    sciSources = InputDatasetField(
+        doc="The catalog of science sources.",
+        name="src",
+        storageClass="SourceCatalog",
+        dimensions={"Instrument", "Exposure", "Detector"},
+    )
+    diaSources = InputDatasetField(
+        doc="The catalog of DIASources.",
+        name="deepDiff_diaSrc",
+        nameTemplate="{coaddName}Diff_diaSrc",
+        storageClass="SourceCatalog",
+        dimensions={"Instrument", "Exposure", "Detector"},
+    )
+
+
+@register("fracDiaSourcesToSciSources")
+class FractionDiaSourcesToSciSourcesMetricTask(MetricTask):
+    """Task that computes the ratio of difference image sources to science
+    sources in an image, visit, etc..
+    """
+    _DefaultName = "fracDiaSourcesToSciSources"
+    ConfigClass = FractionDiaSourcesToSciSourcesMetricConfig
+
+    def run(self, sciSources, diaSources):
+        """Compute the ratio of DIASources to science sources.
+
+        Parameters
+        ----------
+        sciSources : iterable of `lsst.afw.table.SourceCatalog`
+            A collection of science source catalogs, one for each unit of
+            processing to be incorporated into this metric. Its elements may
+            be `None` to represent missing data.
+        diaSources : iterable of `lsst.afw.table.SourceCatalog`
+            A collection of difference imaging catalogs similar to
+            ``sciSources``.
+
+        Returns
+        -------
+        result : `lsst.pipe.base.Struct`
+            A `~lsst.pipe.base.Struct` containing the following component:
+
+            ``measurement``
+                the ratio (`lsst.verify.Measurement` or `None`)
+        """
+        nSciSources = 0
+        nDiaSources = 0
+        inputData = False
+
+        for sciCatalog, diaCatalog in zip(sciSources, diaSources):
+            if diaCatalog is not None and sciCatalog is not None:
+                nSciSources += len(sciCatalog)
+                nDiaSources += len(diaCatalog)
+                inputData = True
+
+        if inputData:
+            metricName = self.getOutputMetricName(self.config)
+            if nSciSources <= 0.0 or nDiaSources <= 0.0:
```

Comment:
I'm not sure what the right behavior here should be if there are zero science sources, but returning 0 seems not quite right from a purely mathematical perspective.  Maybe return `None`?  I'm not sure `NaN` is more informative.

Reply:
FWIW, I copied the behavior from the original [fracUpdatedDiaObjects metric](https://github.com/lsst/ap_verify/blob/master/python/lsst/ap/verify/measurements/association.py#L132-L133).

The current `MetricTask` convention is that `None` means "not enough data" (e.g., the pipeline didn't run difference imaging); having both source catalogs, and them simply being empty, is not the same thing.

On the other hand, maybe `NaN` or `Inf` are acceptable values, since we've seen that it's also perfectly valid to measure `nSciSources == 0` and `nDiaSources > 0`. Though we should check with @afausti to make sure SQuaSH won't choke on it...

Reply:
Or I suppose we could raise `MetricComputationError`, though I'm not sure division by 0 counts as an "algorithmic error preventing calculation of the metric".

Reply:
I definitely don't think this is an error.  It's not clear to me that we need to preserve the semantic distinction between "no diffim" and "nothing in the source catalog", but I agree that if SQuaSH can handle `NaN` let's try that.  `Inf` just seems to be asking for trouble in the plots.

Reply:
@ebellm @kfindeisen SQuaSH will store that and filter out the `NaN`s when displaying things, so it will be fine. However, it would be nice to sanitize the JSON. As pointed out in DM-13179 according to JSON [RFC 4627](https://www.ietf.org/rfc/rfc4627.txt) `NaN`s are not permitted.

Reply:
It sounds as if you want us to use {{null}} instead, is that right?

Reply:
I think sanitizing the JSON must not be the responsibility of the `MetricTask` framework, since users can (and almost certainly will) create and store metrics without it. But it sounds like DM-13179 plans to tackle the root cause anyway.

@ebellm to me the distinction seems very clear: in the former case you don't have the data for the metric to be applicable, in the latter case you *do* have the data, you just don't like what it's telling you. But maybe this is a design issue we should discuss more generally, since I've already found the current policy hard to document.

Reply:
Following offline discussion with @ebellm, this and the tasks for other fractional metrics will raise `MetricComputationError` when the denominator is 0. The higher-level framework will handle things gracefully.